### ISSUE TITLE
Is there any way to refresh the ClientAssertion?

### ISSUE THREAD

**Issue by joshuapetryk** [AzureAD/microsoft-authentication-library-for-dotnet#2184]
Hi - not sure what issue template to use here, please let me know if you'd like me to update. 
I have some odd behavior that I'm not understanding.

I am using the OAuth Auth Code flow and my builder code looks like so:

```
var app = ConfidentialClientApplicationBuilder.Create(clientId)
                .WithClientAssertion(assertion)
                .WithAuthority(AzureCloudInstance.AzurePublic, AadAuthorityAudience.AzureAdMyOrg)
                .WithTenantId(tenantId);
```

I am building the claims like so:

 ```
            string aud = $"https://login.microsoftonline.com/{tenantId}/v2.0";

            const uint JwtToAadLifetimeInSeconds = 60 * 10; // Ten minutes
            DateTime validFrom = DateTime.UtcNow;
            var nbf = ConvertToTimeT(validFrom);
            var exp = ConvertToTimeT(validFrom + TimeSpan.FromSeconds(JwtToAadLifetimeInSeconds));

            return new Dictionary<string, string>()
            {
                { "aud", aud },
                { "exp", exp.ToString() },
                { "iss", clientId },
                { "jti", Guid.NewGuid().ToString() },
                { "nbf", nbf.ToString() },
                { "sub", clientId }
            };
```

We build and persist one client per user session. If a user's session is long running (hours) we start getting the following error message from AAD: "AADSTS700024: Client assertion is not within its valid time range"

I believe this is because `WithClientAssertion` is called during client build time, but the exp claim expires in 10 minutes and we do not build a new client unless the user's session is ended.

Is there a way to update the ClientAssertion without having to rebuild the client again?

**Comment by jmprieur** [comment]
@joshuapetryk : don't you want to use .WithClientAssertion on each call?

**Comment by bgavrilMS** [comment]
@jmprieur - we do not expose `.WithClientAsserttion` on the Acquire* APIs. Do you think we should add this API?

![image](https://user-images.githubusercontent.com/12273384/99064083-5c167180-259d-11eb-91eb-78fd1a6de6db.png)

@joshuapetryk - if you let MSAL create the client assertions from a certificate (which you cannot do I think), we also use [10 min expiration](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/851479f3e55d6ce26aa27c839ce93dc7aaf087ea/src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs#L35) and we reuse the same assertion for [5 minutes](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/851479f3e55d6ce26aa27c839ce93dc7aaf087ea/src/client/Microsoft.Identity.Client/Internal/Requests/ClientCredentialHelper.cs#L37). I think at the moment you'd need to keep track of expiration yourself and re-create the `ConfidentialClientApplication`, as we do not allow a way to re-specify it?

@jmprieur - I see a few ways out of this: 

- add a `.WithClientAssertion(() => string)` i.e. with a callback asking for the assertion
- add `.WithClientAssertion()` to the Acquire* APIs
- add an `UpdateClientAssertion` on the CCA
- didn't we allow in ADAL the possibility for the user to combine using a cert + some extra claims? Most of the time people just need to send some well-known claims and not handle certs on their own.

Personally I think option 1 and 4 would help here.

**Comment by joshuapetryk** [comment]
Thanks @bgavrilMS for confirming this behavior.

Some background: The reason I am forced to use WithClientAssertion is because we are running .Net Framework 4.5 and WithCertificate was throwing exceptions (don't recall which one exactly).

My personal preference would be passing a callback (`.WithClientAssertion(() => string)`) that would be invoked every time when generating the signature.

If I get some time here I will look into making a PR for this enhancement.

**Comment by jmprieur** [comment]
@joshuapetryk thanks for this heads-up.
@bgavrilMS @jennyf19 @trwalke @neha-bhargava @henrik-me, I agree that having a new override of `.WithClientAssertion` on the confidential client application builder taking a delegate to provide the signed assertion would be good. This way the application could even choose to not recompute the client assertion until it expires (as this is a costly operation)

```CSharp
var app = ConfidentialClientApplicationBuilder.Create(clientId)
                .WithClientAssertion(() => { /* compute the client assertion */ })
                .Build();
```

**Comment by jmprieur** [comment]
@joshuapetryk meanwhile, the work-around is to recreate the confidential client application for each request. Is it an issue for you?

**Comment by pmaytak** [comment]
This is included in [MSAL 4.25.0 release](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/releases/tag/4.25.0).

cc: @joshuapetryk
